trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryMetadataDefault2D.java

```diff
@@ -0,0 +1,129 @@
+				.contains(MemoryModelDeclarations.ULTIMATE_LENGTH)) {
+			metaDataDeclarations.add(constructLengthArrayDeclaration());
+		}
+
+		if (requiredFeatures.getRequiredMemoryStructureDeclarations()
+				.contains(MemoryModelDeclarations.ULTIMATE_VALID)) {
+			metaDataDeclarations.add(constructValidArrayDeclaration());
+		}
+
+		if (requiredFeatures.getRequiredMemoryStructureDeclarations()
+				.contains(MemoryModelDeclarations.ULTIMATE_STACK_HEAP_BARRIER)) {
+			metaDataDeclarations.add(constructStackHeapBarrierConstant());
+		}
+
+		return metaDataDeclarations;
+	}
+
+	/**
+	 * Constructs the declaration of the length array, tracking the length of each memory block.
+	 *
+	 * @return The declaration.
+	 */
+	private VariableDeclaration constructLengthArrayDeclaration() {
+		// var #length : [int]int;
+		final ILocation ignoreLoc = LocationFactory.createIgnoreCLocation();
+		final ASTType pointerComponentType =
+				mTypeHandler.cType2AstType(ignoreLoc, mExpressionTranslation.getCTypeOfPointerComponents());
+		final BoogieType boogieType =
+				BoogieType.createArrayType(0, new BoogieType[] { (BoogieType) pointerComponentType.getBoogieType() },
+						(BoogieType) pointerComponentType.getBoogieType());
+		final ASTType lengthType = new ArrayType(ignoreLoc, boogieType, new String[0],
+				new ASTType[] { pointerComponentType }, pointerComponentType);
+		final VarList vlL =
+				new VarList(ignoreLoc, new String[] { MemoryModelDeclarations.ULTIMATE_LENGTH.getName() }, lengthType);
+		return new VariableDeclaration(ignoreLoc, new Attribute[0], new VarList[] { vlL });
+	}
+
+	/**
+	 * Constructs the declaration of the valid array, tracking if a memory block is allocated.
+	 *
+	 * @return The declaration.
+	 */
+	private VariableDeclaration constructValidArrayDeclaration() {
+		// var #valid : [int]bool;
+		final ILocation ignoreLoc = LocationFactory.createIgnoreCLocation();
+		final ASTType pointerComponentType =
+				mTypeHandler.cType2AstType(ignoreLoc, mExpressionTranslation.getCTypeOfPointerComponents());
+		final BoogieType boogieType =
+				BoogieType.createArrayType(0, new BoogieType[] { (BoogieType) pointerComponentType.getBoogieType() },
+						(BoogieType) mBooleanArrayHelper.constructBoolReplacementType().getBoogieType());
+		final ASTType validType = new ArrayType(ignoreLoc, boogieType, new String[0],
+				new ASTType[] { pointerComponentType }, mBooleanArrayHelper.constructBoolReplacementType());
+		final VarList vlV =
+				new VarList(ignoreLoc, new String[] { MemoryModelDeclarations.ULTIMATE_VALID.getName() }, validType);
+		return new VariableDeclaration(ignoreLoc, new Attribute[0], new VarList[] { vlV });
+	}
+
+	@Override
+	public List<MemoryModelDeclarations> getMetaDataDeclarations() {
+		return List.of(MemoryModelDeclarations.ULTIMATE_VALID, MemoryModelDeclarations.ULTIMATE_LENGTH);
+	}
+
+	/**
+	 * Returns the #valid array expression.
+	 *
+	 * @return The expression.
+	 */
+	@SuppressWarnings("static-method")
```

Comment:
Again: Don't suppress any warnings (usually), so we can get the warning in Sonar or Eclipse and potentially fix them.